include/tensor.tcc

```diff
@@ -37,57 +38,79 @@ Tensor<T>::Tensor(const std::string& name,
                   const MemoryLayout mem_layout) :
                   TensorBase(name, data, dims, type, mem_layout)
 {
-    this->_set_tensor_data(data, dims, mem_layout);
+    _set_tensor_data(data, dims, mem_layout);
 }
 
+// Tensor copy constructor
 template <class T>
 Tensor<T>::Tensor(const Tensor<T>& tensor) : TensorBase(tensor)
 {
-    this->_set_tensor_data(tensor._data, tensor._dims,
+    // Check for self-copy
+    if (&tensor == this)
+        return;
+
+    _set_tensor_data(tensor._data, tensor._dims,
                            MemoryLayout::contiguous);
-    this->_c_mem_views = tensor._c_mem_views;
-    this->_f_mem_views = tensor._f_mem_views;
+    _c_mem_views = tensor._c_mem_views;
+    _f_mem_views = tensor._f_mem_views;
 }
 
+// Tensor move constructor
 template <class T>
 Tensor<T>::Tensor(Tensor<T>&& tensor) : TensorBase(std::move(tensor))
 {
-    this->_c_mem_views = std::move(tensor._c_mem_views);
-    this->_f_mem_views = std::move(tensor._f_mem_views);
+    _c_mem_views = std::move(tensor._c_mem_views);
+    _f_mem_views = std::move(tensor._f_mem_views);
 }
 
+// Tensor copy assignment operator
 template <class T>
 Tensor<T>& Tensor<T>::operator=(const Tensor<T>& tensor)
 {
-    if(this!=&tensor) {
-        this->TensorBase::operator=(tensor);
-        this->_set_tensor_data(tensor._data, tensor._dims,
-                               MemoryLayout::contiguous);
-        this->_c_mem_views = tensor._c_mem_views;
-        this->_f_mem_views = tensor._f_mem_views;
-    }
+    // Check for self assignment
+    if (this == &tensor)
+        return *this;
+
+    // Deep copy tensor data
+    TensorBase::operator=(tensor);
+    _set_tensor_data(tensor._data, tensor._dims,
+                            MemoryLayout::contiguous);
+    _c_mem_views = tensor._c_mem_views;
+    _f_mem_views = tensor._f_mem_views;
+
+    // Done
     return *this;
 }
 
+// Tensor move assignment operator
 template <class T>
 Tensor<T>& Tensor<T>::operator=(Tensor<T>&& tensor)
 {
-    if(this!=&tensor) {
-        this->TensorBase::operator=(std::move(tensor));
-        this->_c_mem_views = std::move(tensor._c_mem_views);
-        this->_f_mem_views = std::move(tensor._f_mem_views);
-    }
+    // Check for self-move
+    if (this == &tensor)
+        return *this;
+
+    // Move data
+    TensorBase::operator=(std::move(tensor));
+    _c_mem_views = std::move(tensor._c_mem_views);
+    _f_mem_views = std::move(tensor._f_mem_views);
+
+    // Done
     return *this;
 }
 
+// Deep copy operator
 template <class T>
 TensorBase* Tensor<T>::clone()
 {
     Tensor<T>* new_tensor = new Tensor<T>(*this);
+    if (new_tensor == NULL)
```

Comment:
Isn't the standard that ``new`` will throw ``bad_alloc`` unless ``nothrow`` is passed in?  I think programming to non-standard behavior is going to be a pain to enforce and inconsistent.  If there is an example of a compiler not doing standard behavior, we should put this is a wrapper function or something that is easier to maintain and enforce.

Reply:
It's really hard for me to let go of this one because of how badly I was burned by it, but to be fair the failure I saw was over a decade ago, so it's probably time for me to move on...